File: 프로그래머스/2/42584. 주식가격/주식가격.java
import java.util.*;
class Solution {
    public int[] solution(int[] prices) {
        int[] answer = new int[prices.length];
        Arrays.fill(answer, 0);
        
        Deque<Integer> pricesStack = new ArrayDeque<>();
        
        for(int i=0; i<prices.length; i++){
            answer[i] = prices.length-1-i;
            while(true){
                if(pricesStack.isEmpty()){
                    // 0: index, 1: day
                    pricesStack.push(i);
                    break;
                }
                int currentPriceIndex = pricesStack.peek();
                if(prices[i]<prices[currentPriceIndex]){
                    answer[currentPriceIndex] = i-currentPriceIndex;
                    pricesStack.pop();
                }else{
                    pricesStack.push(i);               
                    break;
                }
            }
        }
        // System.out.println(pricesStack);
        
        return answer;
    }
}

File: 프로그래머스/2/42584. 주식가격/주식가격.js
function solution(prices) {
    const n = prices.length;
    const answer = new Array(n).fill(0);
    const stack = [];  // (인덱스)만 저장

    for (let i = 0; i < n; i++) {
        while (stack.length > 0 && prices[i] < prices[stack[stack.length - 1]]) {
            let j = stack.pop();
            answer[j] = i - j;  // 현재 인덱스에서 이전 가격이 유지된 기간
        }
        stack.push(i);
    }

    // 끝까지 가격이 떨어지지 않은 요소 처리
    while (stack.length > 0) {
        let j = stack.pop();
        answer[j] = n - 1 - j;
    }

    return answer;
}
